containers/EmpSponsors.ts

```diff
@@ -0,0 +1,81 @@
+import { createContainer } from "unstated-next";
+import { useState, useEffect } from "react";
+import { BigNumber, Event } from "ethers";
+
+import Connection from "./Connection";
+import EmpContract from "./EmpContract";
+
+interface PositionState {
+  tokensOutstanding: BigNumber | null;
+  lockedCollateral: BigNumber | null;
+}
+
+interface SponsorDictionary {
+  [address: string]: PositionState;
+}
+
+interface EmpDictionary {
+  [address: string]: SponsorDictionary;
+}
+
+const useEmpSponsors = () => {
+  const { block$ } = Connection.useContainer();
+  const { contract: emp } = EmpContract.useContainer();
+
+  const [activePositions, setActivePositions] = useState<EmpDictionary>({});
+
+  // get position information about every sponsor that has ever created a position.
+  const querySponsors = async () => {
+    if (emp) {
+      const newSponsorFilter = emp.filters.NewSponsor(null);
+      const newSponsorEvents = await emp.queryFilter(newSponsorFilter);
+
+      // Map all active sponsor information to each EMP.
+      if (!activePositions[emp.address]) {
+        activePositions[emp.address] = {};
+      }
```

Comment:
Avoid direct mutation of state variables.

If you need to do things imperatively, make a localized variable for that purpose. e.g.

```ts
let newPositions = { ...activePositions }  // immutably clone object

// perform imperative mutations
if (!newPositions[emp.address]) {
   newPositions[emp.address] = {}
}
```

Reply:
Good call, I actually clone the `activePositions` mapping later with `updatedPositions` but I should do this earlier before I want to modify the mapping at all.